# Éléments nécessaires au TP 1 (chiffrement DES)

F. Goichot, février 2023

Ce document contient :
* les "constantes de DES", c'est-à-dire les différents tableaux définissant les transformations qui interviennent dans le chiffrement d'un message
* un test global et son résultat, pour contrôler votre propre travail
* un test intermédiaire et son résultat, sur le premier tour de chiffrement

## Les constantes de DES 

Source : https://fr.wikipedia.org/wiki/Constantes_du_DES

In [1190]:
import numpy as np
from random import randint

### 1. La permutation initiale $\sigma_0$

In [1191]:
sigma = np.array([58,50,42,34,26,18,10,2,60,52,44,36,28,20,12,4,62,54,46,38,30,22,14,6,64,56,48,40,32,24,16,8,
        57,49,41,33,25,17,9,1,59,51,43,35,27,19,11,3,61,53,45,37,29,21,13,5,63,55,47,39,31,23,15,7], dtype = int)

### 2. La réciproque

C'est $\sigma_0^5$ puisque $\sigma_0$ est d'ordre 6

In [1192]:
sigma_moins_un = [40, 8, 48, 16, 56, 24, 64, 32, 39, 7, 47, 15, 55, 23, 63, 31, 38, 6, 46, 14,
                  54, 22, 62, 30, 37, 5, 45, 13, 53, 21, 61, 29, 36, 4, 44, 12, 52, 20, 60, 28,
                  35, 3, 43, 11, 51, 19, 59, 27, 34, 2, 42, 10, 50, 18, 58, 26, 33, 1, 41, 9, 49, 17, 57, 25]

### 3. Pour les clés de tours

In [1193]:
# Gauche 
PC1g = np.array([57,49,41,33,25,17,9,
1,58,50,42,34,26,18,
10,2,59,51,43,35,27,
19,11,3,60,52,44,36])
	
# Droite 
PC1d = np.array([63,55,47,39,31,23,15,
7,62,54,46,38,30,22,
14,6,61,53,45,37,29,
21,13,5,28,20,12,4])

In [1194]:
PC2 = np.array([14,17,11,24,1,5,
3,28,15,6,21,10,
23,19,12,4,26,8,
16,7,27,20,13,2,
41,52,31,37,47,55,
30,40,51,45,33,48,
44,49,39,56,34,53,
46,42,50,36,29,32])

### 4. Les fonctions de chiffrement $f_K$

#### L'expansion  $E : \{0,1\}^{32} \rightarrow \{0,1\}^{48}$

In [1195]:
E = np.array([32 ,1 ,2 ,3 ,4 ,5,
4 ,5 ,6 ,7 ,8 ,9,
8 ,9 ,10 ,11 ,12 ,13,
12 ,13 ,14 ,15 ,16 ,17,
16 ,17 ,18 ,19 ,20 ,21,
20 ,21 ,22 ,23 ,24 ,25,
24 ,25 ,26 ,27 ,28 ,29,
28 ,29 ,30 ,31 ,32 ,1 ])

#### La permutation de sortie $P \in S_{32}$

In [1196]:
P  = np.array([16,7,20,21,
29,12,28,17,
1,15,23,26,
5,18,31,10,
2,8,24,14,
32,27,3,9,
19,13,30,6,
22,11,4,25])

#### Les S-boîtes

In [1197]:
S1 = np.array([
[14,4,13,1,2,15,11,8,3,10,6,12,5,9,0,7],
[0,15,7,4,14,2,13,1,10,6,12,11,9,5,3,8],
[4,1,14,8,13,6,2,11,15,12,9,7,3,10,5,0],
[15,12,8,2,4,9,1,7,5,11,3,14,10,0,6,13]])

S2 = np.array([
[15,1,8,14,6,11,3,4,9,7,2,13,12,0,5,10],
[3,13,4,7,15,2,8,14,12,0,1,10,6,9,11,5],
[0,14,7,11,10,4,13,1,5,8,12,6,9,3,2,15],
[13,8,10,1,3,15,4,2,11,6,7,12,0,5,14,9]])

S3 = np.array([
[10,0,9,14,6,3,15,5,1,13,12,7,11,4,2,8],
[13,7,0,9,3,4,6,10,2,8,5,14,12,11,15,1],
[13,6,4,9,8,15,3,0,11,1,2,12,5,10,14,7],
[1,10,13,0,6,9,8,7,4,15,14,3,11,5,2,12]])

S4 = np.array([
[7,13,14,3,0,6,9,10,1,2,8,5,11,12,4,15],
[13,8,11,5,6,15,0,3,4,7,2,12,1,10,14,9],
[10,6,9,0,12,11,7,13,15,1,3,14,5,2,8,4],
[3,15,0,6,10,1,13,8,9,4,5,11,12,7,2,14]])

S5 = np.array([
[2,12,4,1,7,10,11,6,8,5,3,15,13,0,14,9],
[14,11,2,12,4,7,13,1,5,0,15,10,3,9,8,6],
[4,2,1,11,10,13,7,8,15,9,12,5,6,3,0,14],
[11,8,12,7,1,14,2,13,6,15,0,9,10,4,5,3]])

S6 = np.array([
[12,1,10,15,9,2,6,8,0,13,3,4,14,7,5,11],
[10,15,4,2,7,12,9,5,6,1,13,14,0,11,3,8],
[9,14,15,5,2,8,12,3,7,0,4,10,1,13,11,6],
[4,3,2,12,9,5,15,10,11,14,1,7,6,0,8,13]])

S7 = np.array([
[4,11,2,14,15,0,8,13,3,12,9,7,5,10,6,1],
[13,0,11,7,4,9,1,10,14,3,5,12,2,15,8,6],
[1,4,11,13,12,3,7,14,10,15,6,8,0,5,9,2],
[6,11,13,8,1,4,10,7,9,5,0,15,14,2,3,12]])

S8 = np.array([
[13,2,8,4,6,15,11,1,10,9,3,14,5,0,12,7],
[1,15,13,8,10,3,7,4,12,5,6,11,0,14,9,2],
[7,11,4,1,9,12,14,2,0,6,10,13,15,3,5,8],
[2,1,14,7,4,10,8,13,15,12,9,0,3,5,6,11]])

In [1198]:
S = np.array([S1,S2,S3,S4,S5,S6,S7,S8]) ; S[0][0,1]

4

## Test de contrôle


In [1199]:
# Message en clair 
a = np.array([1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0,
       0, 0, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0,
       0, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0])
# soit en hexadécimal
'F00198E6EC84F868'
# Clé de session
b = np.array([1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 1, 1, 0, 1,
       1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1,
       0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1])
# soit en hexadécimal
'DF26B60D7691737F'
# Message crypté
c = np.array([1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0,
       0, 0, 0, 1, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1,
       1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 1, 1, 0, 0])
# soit en hexadécimal
'A61CA055AABCF79C'

'A61CA055AABCF79C'

### Contrôle intermédiaire pour les clés de tour et les fonctions de chiffrement de tour $f_K$

```py
# même message en clair  et même clé que ci-dessus
# la première clé de tour
[0 1 1 1 0 1 0 0]
[1 0 1 0 0 1 0 0]
[1 1 1 1 1 0 0 1]
[1 0 1 0 0 1 0 1]
[1 0 1 1 1 0 1 1]
[1 1 0 1 0 1 1 0]
# soit en hexadécimal
'74A4F9A5BBD6'
# le résultat de f_K1 sur la moitié droite du message en clair (sans permutation)
[1 0 1 0 1 1 0 0]
[1 1 0 1 0 1 0 1]
[1 0 1 1 0 0 1 0]
[0 1 0 0 1 1 0 0]
# soit en hexadécimal
'ACD5B24C'
```

----

# Debut du TP1

## Creations des clés de tours

In [1200]:
class DES_Key():
    key:np.ndarray
    C0:np.ndarray = np.zeros(1)
    D0:np.ndarray = np.zeros(1)
    TourKeys:list[np.ndarray] = []
    def __init__(self,key:np.ndarray):
        if len(key) != 64 : #Verifie que la longueur est bonne
            raise IndexError("La clé ne fait pas la bonne taille")
        for i in range(0,64,8): # vérifie si chaque octet est impair
            if key[i:i+8].sum()%2==0:
                raise ValueError("La clé doit être composée d'octets impairs")
        self.key = key
    def generateC0D0(self):
        #on genere C0 à partir de PC1g et PC1d
        self.C0 = np.zeros(28,int)
        self.D0 = np.zeros(28,int)
        for i,index in enumerate(PC1g) :
            self.C0[i] = self.key[index-1]
        for i,index in enumerate(PC1d) :
            self.D0[i] = self.key[index-1]
    def getC0D0(self):
        if len(self.C0) != 28 or len(self.D0) != 28 :
            self.generateC0D0()
        return self.C0,self.D0
    def getClesdeTour(self)->list[np.ndarray]:
        # génération des clés de tours
        if len(self.TourKeys) != 16 :
            c0,d0 = self.getC0D0()
            c0 = list(c0)
            d0 = list(d0)
            for i in range(1, 17):
                vi = 1 if i in [1, 2, 9, 16] else 2
                ci = c0[vi:] + c0[:vi]
                di = d0[vi:] + d0[:vi]
                c0 = ci
                d0 = di
                valPhi2 = np.array(c0+d0)
                ki = np.array([valPhi2[val-1] for val in PC2],int)
                self.TourKeys.append(ki)
        return self.TourKeys

def afficheOctets(cle,):
    for i in range(0,len(cle),8):
        print(cle[i:i+8])

def afficheHexa(cle,):
    hexe = "0123456789ABCDEF"
    mot = ""
    for i in range(0,len(cle),4):
        compt = 0
        for j,item in enumerate(cle[i:i+4]):
            compt += item*(2**(3-j))
        mot += hexe[compt]
    print(mot)

#--------------------------------------------------------

cle1 = DES_Key(b)
tour1 = cle1.getClesdeTour()
afficheOctets(cle1.key)
afficheHexa(cle1.key)
print()
for item in tour1 :
    afficheOctets(item)
    afficheHexa(item)
    print()

[1 1 0 1 1 1 1 1]
[0 0 1 0 0 1 1 0]
[1 0 1 1 0 1 1 0]
[0 0 0 0 1 1 0 1]
[0 1 1 1 0 1 1 0]
[1 0 0 1 0 0 0 1]
[0 1 1 1 0 0 1 1]
[0 1 1 1 1 1 1 1]
DF26B60D7691737F

[0 1 1 1 0 1 0 0]
[1 0 1 0 0 1 0 0]
[1 1 1 1 1 0 0 1]
[1 0 1 0 0 1 0 1]
[1 0 1 1 1 0 1 1]
[1 1 0 1 0 1 1 0]
74A4F9A5BBD6

[1 0 0 1 1 0 0 0]
[1 1 1 1 1 0 0 1]
[1 1 1 1 0 1 0 0]
[1 1 0 0 1 1 1 1]
[0 0 1 0 0 0 1 1]
[1 0 0 1 1 0 0 1]
98F9F4CF2399

[1 0 0 1 0 1 0 0]
[0 1 1 0 1 1 1 1]
[0 1 1 0 1 0 1 1]
[1 1 1 1 0 0 1 1]
[0 1 1 1 0 0 1 1]
[0 1 0 0 1 1 0 1]
946F6BF3734D

[1 1 1 0 0 0 1 1]
[0 1 1 1 0 1 1 1]
[0 0 1 0 0 1 0 1]
[0 1 1 1 0 0 1 0]
[1 0 0 1 0 0 1 1]
[1 0 1 0 1 1 1 0]
E377257293AE

[1 1 0 0 1 0 0 1]
[1 0 0 1 1 1 1 1]
[1 1 0 0 0 1 0 1]
[1 1 0 1 0 1 0 0]
[0 0 1 1 1 1 0 1]
[1 0 1 0 1 1 1 1]
C99FC5D43DAF

[0 1 0 1 0 0 0 1]
[1 1 1 1 0 0 1 0]
[1 1 1 1 1 0 1 1]
[0 1 1 0 1 1 1 0]
[0 0 1 1 1 0 1 0]
[1 1 1 1 1 0 0 1]
51F2FB6E3AF9

[1 0 1 1 0 1 0 1]
[1 1 0 1 0 1 0 1]
[1 1 1 0 0 0 1 0]
[0 1 1 1 0 0 1 1]
[1 1 1 1 1 0 0 1]
[0 1 1 1 1 0 1 1

En comparant nos resultat avec l'exemple, on voit qu'on a les mêmes resultats

## Créations des fonction pour $f_K$ et tests

In [1201]:
def split(msg):
    return msg[:32],msg[32:]

def Expansion(demimot:np.ndarray)->np.ndarray:
    ret = np.zeros(48,int)
    for i in range(0,48):
        ret[i] = demimot[E[i]-1]
    return ret

def XOR(a,b):
    if len(a) != len(b) :
        raise IndexError("a et b ne sont pas de la meme taille")
    ret = np.zeros(len(a),int)
    for i in range (0,len(a)):
        if a[i]==1 and b[i] == 1 :
            ret[i] = 0
        else :
            ret[i] = a[i] + b[i]
    return ret

def perm(mot,permu)->np.ndarray:
    #permet d'effectuer n'importe quelle permutation
    ret = np.zeros(len(mot),int)
    for i in range (0,len(mot)):
        ret[i] = mot[permu[i]-1]
    return ret


perminitial = a #perm(a,sigma)
afficheOctets(perminitial)
print()
l0,r0 = split(perminitial)
afficheOctets(r0)
print()
Er0 = Expansion(r0)
afficheOctets(Er0)
print()
afficheOctets(cle1.getClesdeTour()[0])
print()
xored = XOR(Er0,cle1.getClesdeTour()[0])
afficheOctets(xored)

[1 1 1 1 0 0 0 0]
[0 0 0 0 0 0 0 1]
[1 0 0 1 1 0 0 0]
[1 1 1 0 0 1 1 0]
[1 1 1 0 1 1 0 0]
[1 0 0 0 0 1 0 0]
[1 1 1 1 1 0 0 0]
[0 1 1 0 1 0 0 0]

[1 1 1 0 1 1 0 0]
[1 0 0 0 0 1 0 0]
[1 1 1 1 1 0 0 0]
[0 1 1 0 1 0 0 0]

[0 1 1 1 0 1 0 1]
[1 0 0 1 0 1 0 0]
[0 0 0 0 1 0 0 1]
[0 1 1 1 1 1 1 1]
[0 0 0 0 0 0 1 1]
[0 1 0 1 0 0 0 1]

[0 1 1 1 0 1 0 0]
[1 0 1 0 0 1 0 0]
[1 1 1 1 1 0 0 1]
[1 0 1 0 0 1 0 1]
[1 0 1 1 1 0 1 1]
[1 1 0 1 0 1 1 0]

[0 0 0 0 0 0 0 1]
[0 0 1 1 0 0 0 0]
[1 1 1 1 0 0 0 0]
[1 1 0 1 1 0 1 0]
[1 0 1 1 1 0 0 0]
[1 0 0 0 0 1 1 1]


In [1202]:
def calc(mot6):
    #retourne b1b6 et B2B3B4B5
    return mot6[0]*2+mot6[5],mot6[1]*8+mot6[2]*4+mot6[3]*2+mot6[4]

def int_to_bin(value):
    #conversion en binaire
    l = []
    while value != 0 :
        l.append(value%2)
        value = value//2
    while len(l) < 4:
        l.append(0)
    return l[::-1]

def s_boite(demimot):
    #passage du demi mot à travers les s boites
    res = []
    for i in range (0,8):
        mot6 = demimot[i*6:(i+1)*6]
        a,b = calc(mot6)
        indsboite = S[i][a,b]
        res += int_to_bin(indsboite)
    return np.array(res)

result = perm(s_boite(xored),P)
afficheOctets(result)
afficheHexa(result)

[1 0 1 0 1 1 0 0]
[1 1 0 1 0 1 0 1]
[1 0 1 1 0 0 1 0]
[0 1 0 0 1 1 0 0]
ACD5B24C


On retrouve bien le resultat donné dans l'exemple plus haut

## Creation de la fonction de chiffrement `DES(k,p)`

In [1203]:
def f_k(key:DES_Key,demimot,tour:int):
    #fonctions Fk
    exp = Expansion(demimot)
    xored = XOR(exp,key.getClesdeTour()[tour])
    return perm(s_boite(xored),P)

def DES(k:DES_Key,p:np.ndarray):
    motp = perm(p,sigma)
    l0,r0 = split(motp)
    for i in range(0,16):
        nl = r0
        nr = XOR(l0,f_k(k,r0,i))
        l0 = nl
        r0 = nr
    premotf = np.concatenate((nr,nl))
    return perm(premotf,sigma_moins_un)

dessed = DES(cle1,a)

afficheOctets(dessed)
afficheHexa(dessed)

dessed == c


[1 0 1 0 0 1 1 0]
[0 0 0 1 1 1 0 0]
[1 0 1 0 0 0 0 0]
[0 1 0 1 0 1 0 1]
[1 0 1 0 1 0 1 0]
[1 0 1 1 1 1 0 0]
[1 1 1 1 0 1 1 1]
[1 0 0 1 1 1 0 0]
A61CA055AABCF79C


array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True])

## Creation de la fonction de déchiffrement `unDES(k,p)`

On effectue alors les mêmes opérations avec les tours dans le sens inverse

In [1204]:
def unDES(k:DES_Key,p:np.ndarray):
    motp = perm(p,sigma)
    r0,l0 = split(motp)
    for i in range(0,16):
        nr = l0
        nl = XOR(r0,f_k(k,l0,15-i))
        l0 = nl
        r0 = nr
        
    premotf = np.concatenate((nl,nr))
    return perm(premotf,sigma_moins_un)

undessed = unDES(cle1,dessed)

afficheOctets(undessed)
afficheHexa(undessed)

undessed == a

[1 1 1 1 0 0 0 0]
[0 0 0 0 0 0 0 1]
[1 0 0 1 1 0 0 0]
[1 1 1 0 0 1 1 0]
[1 1 1 0 1 1 0 0]
[1 0 0 0 0 1 0 0]
[1 1 1 1 1 0 0 0]
[0 1 1 0 1 0 0 0]
F00198E6EC84F868


array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True])

Nous retombons alors sur le message en clair

## Analyse Differentielle

In [1205]:
def swapbit(a:np.ndarray,b:int):
    if b > len(a) :
        raise ValueError("b trop grand pour a")
    ap = np.copy(a)
    ap[b] = ap[b]^1
    return ap


modified = swapbit(a,randint(0,63))

desa = DES(cle1,a)
desm = DES(cle1,modified)

diff = np.array(desa == desm,int)

aff = ['I','D']
res = ['']*64
for i in range(0,64):
    res[i] = aff[diff[i]]
    

afficheOctets(res)

print(f"nombre de bit differents : {diff.sum()}")

['I', 'I', 'D', 'I', 'I', 'D', 'I', 'D']
['I', 'I', 'D', 'I', 'D', 'I', 'I', 'I']
['D', 'D', 'D', 'I', 'I', 'D', 'D', 'I']
['I', 'I', 'I', 'I', 'D', 'I', 'D', 'D']
['I', 'I', 'D', 'I', 'D', 'I', 'D', 'I']
['I', 'D', 'I', 'I', 'D', 'I', 'D', 'I']
['D', 'I', 'D', 'D', 'D', 'I', 'I', 'I']
['I', 'I', 'I', 'D', 'D', 'I', 'D', 'I']
nombre de bit differents : 26


Les bits identiques sont marqué `I`

Dans le bloc suivant, on teste en modifiant tous les bits du mot donné, on se rend compte qu'on a seulement la moitié des bits qui sont bons, ce qui correspond statistiquement à ce qu'on aurait aléatoirement.

In [1206]:
moyenne = 0.0
for i in range(0,64):
    modified = swapbit(a,i)

    desa = DES(cle1,a)
    desm = DES(cle1,modified)

    diff = np.array(desa == desm,int)

    moyenne += diff.sum()

moyenne /= 64


print(f"nombre de bit moyen differents : {moyenne}")

nombre de bit moyen differents : 31.328125


On a donc en moyenne $31.328125$ bit differents en changeant un bit c'est a dire quasiment la moitié. puisque que on a une chance sur deux qu'un bit choisi aléatoirement soit identique, on peut donc dire que l'algorithme DES est très resistant à l'analyse differentielle